### ISSUE TITLE
Using Rank Propensity Scores

### ISSUE THREAD

**Issue by eggie5** [tensorflow/ranking#190]
If I had propensity scores for the first 7 ranks, like in this image: 

![https://ww1.sinaimg.cn/large/006tNc79gy1g58r0twloxj31400u04qt.jpg](https://ww1.sinaimg.cn/large/006tNc79gy1g58r0twloxj31400u04qt.jpg)

How do I use them in TFR? Looking for a simple example of the API (please not the Keras API -- not ready for that yet 😄 ).

**Comment by nishathussain** [comment]
As per my understanding: one has to change the loss function as 
propensities could be 
1.`position-bais` (the propensity of having click at the clicked position )
2.`query-document` pair. (the propensity of a document being clicked for a given query)

1. With only `position-bais`: (if you want to implement positional propensity only)
new_loss = old_loss/propensity_position

2. `query-doc`+ `position-bais`: (both position and query document pair propensity)
new_loss = old_loss/( propensity_position * propensity_query_doc )

**Comment by xuanhuiwang** [comment]
@eggie5, This can be done in TFR. You need to have a feature in your input data like "doc_weight". Then you can follow this commit to enable weighted loss and weighted metrics: https://github.com/tensorflow/ranking/commit/029fffb834fefb6727506ff3db3e9ec262223fb9.

**Comment by eggie5** [comment]
@xuanhuiwang is there a notion of per-document weighting vs per-list weighting. What if I only have global weights for the ranks. Seems like the commit you shared assumes per-document weights, right?

**Comment by xuanhuiwang** [comment]
Our current code supports both per-list weights and per-document weights in the same manner. 

If you have a weight per rank, this can be cast to a per-document weight by looking up the weight using the document presented rank. Note the rank here is the one when the document was displayed to a user. So it is static in the training data and not the dynamic rank that is determined by the logits during training.

**Comment by xuanhuiwang** [comment]
Close this for now. Feel free to reopen!
